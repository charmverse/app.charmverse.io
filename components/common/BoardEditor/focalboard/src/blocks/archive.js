"use strict";
exports.__esModule = true;
exports.ArchiveUtils = void 0;
var ArchiveUtils = /** @class */ (function () {
    function ArchiveUtils() {
    }
    ArchiveUtils.buildBlockArchive = function (blocks) {
        var header = {
            version: 1,
            date: Date.now()
        };
        var headerString = JSON.stringify(header);
        var content = "".concat(headerString, "\n");
        for (var _i = 0, blocks_1 = blocks; _i < blocks_1.length; _i++) {
            var block = blocks_1[_i];
            var line = {
                type: 'block',
                data: block
            };
            var lineString = JSON.stringify(line);
            content += lineString;
            content += '\n';
        }
        return content;
    };
    ArchiveUtils.parseBlockArchive = function (contents) {
        var blocks = [];
        var allLineStrings = contents.split('\n');
        if (allLineStrings.length >= 2) {
            var headerString = allLineStrings[0];
            var header = JSON.parse(headerString);
            if (header.date && header.version >= 1) {
                var lineStrings = allLineStrings.slice(1);
                var lineNum = 2;
                for (var _i = 0, lineStrings_1 = lineStrings; _i < lineStrings_1.length; _i++) {
                    var lineString = lineStrings_1[_i];
                    if (!lineString) {
                        // Ignore empty lines, e.g. last line
                        continue;
                    }
                    var line = JSON.parse(lineString);
                    if (!line || !line.type || !line.data) {
                        throw new Error("ERROR parsing line ".concat(lineNum));
                    }
                    switch (line.type) {
                        case 'block': {
                            var blockLine = line;
                            var block = blockLine.data;
                            blocks.push(block);
                            break;
                        }
                    }
                    lineNum += 1;
                }
            }
            else {
                throw new Error('ERROR parsing header');
            }
        }
        return blocks;
    };
    return ArchiveUtils;
}());
exports.ArchiveUtils = ArchiveUtils;
