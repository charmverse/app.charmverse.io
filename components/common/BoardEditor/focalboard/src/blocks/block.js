"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports.createPatchesFromBlocks = exports.createBlock = exports.contentBlockTypes = exports.blockTypes = void 0;
var difference_1 = require("lodash/difference");
var utils_1 = require("../utils");
var contentBlockTypes = ['text', 'image', 'divider', 'checkbox'];
exports.contentBlockTypes = contentBlockTypes;
var blockTypes = __spreadArray(__spreadArray([], contentBlockTypes, true), ['board', 'view', 'card', 'comment', 'unknown'], false);
exports.blockTypes = blockTypes;
function createBlock(block) {
    var now = Date.now();
    return {
        id: (block === null || block === void 0 ? void 0 : block.id) || utils_1.Utils.createGuid(utils_1.Utils.blockTypeToIDType(block === null || block === void 0 ? void 0 : block.type)),
        schema: 1,
        spaceId: (block === null || block === void 0 ? void 0 : block.spaceId) || '',
        parentId: (block === null || block === void 0 ? void 0 : block.parentId) || '',
        rootId: (block === null || block === void 0 ? void 0 : block.rootId) || '',
        createdBy: (block === null || block === void 0 ? void 0 : block.createdBy) || '',
        updatedBy: (block === null || block === void 0 ? void 0 : block.updatedBy) || '',
        type: (block === null || block === void 0 ? void 0 : block.type) || 'unknown',
        fields: (block === null || block === void 0 ? void 0 : block.fields) ? __assign({}, block === null || block === void 0 ? void 0 : block.fields) : {},
        title: (block === null || block === void 0 ? void 0 : block.title) || '',
        createdAt: (block === null || block === void 0 ? void 0 : block.createdAt) || now,
        updatedAt: (block === null || block === void 0 ? void 0 : block.updatedAt) || now,
        deletedAt: (block === null || block === void 0 ? void 0 : block.deletedAt) || null
    };
}
exports.createBlock = createBlock;
// createPatchesFromBlock creates two BlockPatch instances, one that
// contains the delta to update the block and another one for the undo
// action, in case it happens
function createPatchesFromBlocks(newBlock, oldBlock) {
    var newDeletedFields = (0, difference_1["default"])(Object.keys(newBlock.fields), Object.keys(oldBlock.fields));
    var newUpdatedFields = {};
    var newUpdatedData = {};
    Object.keys(newBlock.fields).forEach(function (val) {
        if (oldBlock.fields[val] !== newBlock.fields[val]) {
            newUpdatedFields[val] = newBlock.fields[val];
        }
    });
    Object.keys(newBlock).forEach(function (val) {
        if (val !== 'fields' && oldBlock[val] !== newBlock[val]) {
            newUpdatedData[val] = newBlock[val];
        }
    });
    var oldDeletedFields = (0, difference_1["default"])(Object.keys(oldBlock.fields), Object.keys(newBlock.fields));
    var oldUpdatedFields = {};
    var oldUpdatedData = {};
    Object.keys(oldBlock.fields).forEach(function (val) {
        if (oldBlock.fields[val] !== newBlock.fields[val]) {
            oldUpdatedFields[val] = oldBlock.fields[val];
        }
    });
    Object.keys(oldBlock).forEach(function (val) {
        if (val !== 'fields' && oldBlock[val] !== newBlock[val]) {
            oldUpdatedData[val] = oldBlock[val];
        }
    });
    return [
        __assign(__assign({}, newUpdatedData), { updatedFields: newUpdatedFields, deletedFields: oldDeletedFields }),
        __assign(__assign({}, oldUpdatedData), { updatedFields: oldUpdatedFields, deletedFields: newDeletedFields })
    ];
}
exports.createPatchesFromBlocks = createPatchesFromBlocks;
