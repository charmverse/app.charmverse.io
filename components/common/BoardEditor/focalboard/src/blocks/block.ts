
import difference from 'lodash/difference';

import { Utils } from '../utils';

const contentBlockTypes = ['text', 'image', 'divider', 'checkbox'] as const;
const blockTypes = [...contentBlockTypes, 'board', 'view', 'card', 'comment', 'unknown'] as const;
type ContentBlockTypes = typeof contentBlockTypes[number]
type BlockTypes = typeof blockTypes[number]

interface BlockPatch {
    spaceId?: string;
    parentId?: string;
    rootId?: string;
    schema?: number;
    type?: BlockTypes;
    title?: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    updatedFields?: Record<string, any>;
    deletedFields?: string[];
    deletedAt?: number;
}

interface Block {
    id: string;
    spaceId: string;
    parentId: string;
    rootId: string;
    createdBy: string;
    updatedBy: string;
    schema: number;
    type: BlockTypes;
    title: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    fields: Record<string, any>;

    createdAt: number;
    updatedAt: number;
    deletedAt: number | null;
}

function createBlock (block?: Partial<Block>): Block {
  const now = Date.now();
  return {
    id: block?.id || Utils.createGuid(Utils.blockTypeToIDType(block?.type)),
    schema: 1,
    spaceId: block?.spaceId || '',
    parentId: block?.parentId || '',
    rootId: block?.rootId || '',
    createdBy: block?.createdBy || '',
    updatedBy: block?.updatedBy || '',
    type: block?.type || 'unknown',
    fields: block?.fields ? { ...block?.fields } : {},
    title: block?.title || '',
    createdAt: block?.createdAt || now,
    updatedAt: block?.updatedAt || now,
    deletedAt: block?.deletedAt || null
  };
}

// createPatchesFromBlock creates two BlockPatch instances, one that
// contains the delta to update the block and another one for the undo
// action, in case it happens
function createPatchesFromBlocks (newBlock: Block, oldBlock: Block): BlockPatch[] {
  const newDeletedFields = difference(Object.keys(newBlock.fields), Object.keys(oldBlock.fields));
  const newUpdatedFields: Record<string, any> = {};
  const newUpdatedData: Record<string, any> = {};
  Object.keys(newBlock.fields).forEach((val) => {
    if (oldBlock.fields[val] !== newBlock.fields[val]) {
      newUpdatedFields[val] = newBlock.fields[val];
    }
  });
  Object.keys(newBlock).forEach((val) => {
    if (val !== 'fields' && (oldBlock as any)[val] !== (newBlock as any)[val]) {
      newUpdatedData[val] = (newBlock as any)[val];
    }
  });

  const oldDeletedFields = difference(Object.keys(oldBlock.fields), Object.keys(newBlock.fields));
  const oldUpdatedFields: Record<string, any> = {};
  const oldUpdatedData: Record<string, any> = {};
  Object.keys(oldBlock.fields).forEach((val) => {
    if (oldBlock.fields[val] !== newBlock.fields[val]) {
      oldUpdatedFields[val] = oldBlock.fields[val];
    }
  });
  Object.keys(oldBlock).forEach((val) => {
    if (val !== 'fields' && (oldBlock as any)[val] !== (newBlock as any)[val]) {
      oldUpdatedData[val] = (oldBlock as any)[val];
    }
  });

  return [
    {
      ...newUpdatedData,
      updatedFields: newUpdatedFields,
      deletedFields: oldDeletedFields
    },
    {
      ...oldUpdatedData,
      updatedFields: oldUpdatedFields,
      deletedFields: newDeletedFields
    }
  ];
}

export type { ContentBlockTypes, BlockTypes, Block, BlockPatch };
export { blockTypes, contentBlockTypes, createBlock, createPatchesFromBlocks };

