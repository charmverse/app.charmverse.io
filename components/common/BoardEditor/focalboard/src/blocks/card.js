"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.createCard = void 0;
var block_1 = require("./block");
/**
 * Returns a focalboard-ready card data stub
 */
function createCard(block) {
    var _a, _b, _c, _d, _e, _f;
    var contentOrder = [];
    var contentIds = (_b = (_a = block === null || block === void 0 ? void 0 : block.fields) === null || _a === void 0 ? void 0 : _a.contentOrder) === null || _b === void 0 ? void 0 : _b.filter(function (id) { return id !== null; });
    if ((contentIds === null || contentIds === void 0 ? void 0 : contentIds.length) > 0) {
        for (var _i = 0, contentIds_1 = contentIds; _i < contentIds_1.length; _i++) {
            var contentId = contentIds_1[_i];
            if (typeof contentId === 'string') {
                contentOrder.push(contentId);
            }
            else {
                contentOrder.push(contentId.slice());
            }
        }
    }
    return __assign(__assign({}, (0, block_1.createBlock)(block)), { type: 'card', fields: {
            icon: ((_c = block === null || block === void 0 ? void 0 : block.fields) === null || _c === void 0 ? void 0 : _c.icon) || '',
            properties: __assign({}, (((_d = block === null || block === void 0 ? void 0 : block.fields) === null || _d === void 0 ? void 0 : _d.properties) || {})),
            contentOrder: contentOrder,
            isTemplate: ((_e = block === null || block === void 0 ? void 0 : block.fields) === null || _e === void 0 ? void 0 : _e.isTemplate) || false,
            headerImage: ((_f = block === null || block === void 0 ? void 0 : block.fields) === null || _f === void 0 ? void 0 : _f.headerImage) || null
        } });
}
exports.createCard = createCard;
