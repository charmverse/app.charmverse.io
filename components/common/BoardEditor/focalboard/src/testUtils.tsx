import type { PageMeta } from '@charmverse/core/pages';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { IntlProvider } from 'react-intl';
import type { Middleware } from 'redux';
import type { MockStoreEnhanced } from 'redux-mock-store';
import configureStore from 'redux-mock-store';
import { v4 } from 'uuid';

import { PagesContext } from 'hooks/usePages';
import type { Block } from 'lib/focalboard/block';
import { pageStubToCreate } from 'testing/generatePageStub';
import { AppThemeProvider } from 'theme/AppThemeProvider';

export const wrapIntl = (children?: React.ReactNode): JSX.Element => (
  <AppThemeProvider>
    <IntlProvider locale='en'>{children}</IntlProvider>
  </AppThemeProvider>
);
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
  return <DndProvider backend={HTML5Backend}>{wrapIntl(children)}</DndProvider>;
};

export const wrapPagesProvider = (cardId: string | string[], children: React.ReactNode) => {
  const pages: Record<string, PageMeta> = {};

  const cardIds = Array.isArray(cardId) ? cardId : [cardId];

  cardIds.forEach((_cardId) => {
    pages[_cardId] = pageStubToCreate({ spaceId: v4(), createdBy: v4(), id: _cardId, path: _cardId }) as PageMeta;
  });

  return (
    <PagesContext.Provider
      value={
        {
          pages
        } as any
      }
    >
      {children}
    </PagesContext.Provider>
  );
};

export function mockDOM(): void {
  window.focus = jest.fn();
  document.createRange = () => {
    const range = new Range();
    range.getBoundingClientRect = jest.fn();
    range.getClientRects = () => {
      return {
        item: () => null,
        length: 0,
        [Symbol.iterator]: jest.fn()
      };
    };
    return range;
  };
}
export function mockMatchMedia(result: { matches: boolean }): void {
  // We check if system preference is dark or light theme.
  // This is required to provide it's definition since
  // window.matchMedia doesn't exist in Jest.
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => {
      return result;

      // return ({
      //     matches: true,
      // })
    })
  });
}

export function mockStateStore(middleware: Middleware[], state: unknown): MockStoreEnhanced<unknown, unknown> {
  const mockStore = configureStore(middleware);
  return mockStore(state);
}

export type BlocksById<BlockType> = { [key: string]: BlockType };

export function blocksById<BlockType extends Block>(blocks: BlockType[]): BlocksById<BlockType> {
  return blocks.reduce((res, block) => {
    res[block.id] = block;
    return res;
  }, {} as BlocksById<BlockType>);
}
