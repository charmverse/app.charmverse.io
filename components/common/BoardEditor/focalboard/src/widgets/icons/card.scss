.CardIcon {
    stroke-width: 6px;
}
