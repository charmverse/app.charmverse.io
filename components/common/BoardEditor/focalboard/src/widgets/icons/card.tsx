
import React from 'react';

export default function CardIcon (): JSX.Element {
  return (
    <svg
      xmlns='http://www.w3.org/2000/svg'
      className='CardIcon Icon'
      viewBox='0 0 100 100'
    >
      <rect
        x='20'
        y='30'
        width='60'
        height='40'
        rx='3'
        ry='3'
      />
    </svg>
  );
}
