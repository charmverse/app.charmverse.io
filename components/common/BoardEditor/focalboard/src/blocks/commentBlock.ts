import type { PageContent } from 'models';

import type { Block } from './block';
import { createBlock } from './block';

type CommentBlock = Block & {
  type: 'comment';
  fields: { content?: PageContent }; // note: the 'text' value is saved as block.title
}

function createCommentBlock (block?: Block): CommentBlock {
  return {
    ...createBlock(block),
    type: 'comment'
  };
}

export { createCommentBlock };
export type { CommentBlock };
