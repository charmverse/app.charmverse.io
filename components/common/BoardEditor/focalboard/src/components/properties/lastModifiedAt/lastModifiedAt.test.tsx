
import React from 'react';
import { render } from '@testing-library/react';

import { wrapIntl } from '../../../testUtils';

import { createCard } from '../../../blocks/card';
import { createCommentBlock } from '../../../blocks/commentBlock';

import LastModifiedAt from './lastModifiedAt';

describe('componnets/properties/lastModifiedAt', () => {
  test('should match snapshot', () => {
    const card = createCard();
    card.id = 'card-id-1';

    const component = wrapIntl(
      <LastModifiedAt
        updatedAt=''
      />
    );

    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});
