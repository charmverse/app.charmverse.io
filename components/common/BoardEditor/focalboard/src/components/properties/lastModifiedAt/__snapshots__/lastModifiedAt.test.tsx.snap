// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componnets/properties/lastModifiedAt should match snapshot 1`] = `
<div>
  <div
    class="LastModifiedAt octo-propertyvalue readonly"
  >
    June 15, 2021, 4:22 PM
  </div>
</div>
`;
